### PR TITLE
Rewrite the arduino starter kit in johnny-five

### PR DESCRIPTION
**Pull request by warycat**
Add more examples to the popular arduino starter kit.

### CODE REVIEW

**Comment by rwaldron** [comment]
This is awesome! Before we can merge this, it needs a bit of work. I will spend some time identifying "todo" items and will post here

**Comment by warycat** [comment]
Thanks Rick, that will be great for me to learn.

**Comment by rwaldron** [comment]
This will be a mixed bag of tasks, I will try my best to keep it orderly...

- Move all of the JS files to `eg/`, renamed with some useful prefix (just to avoid duplicate collisions). This can easily be done mechanically, eg. with a node.js script or similar. "arduino-kit-" seems good to me. 
- Run `grunt` and identify all the lint nits that have to be addressed
- Add fritzing diagrams to [docs/breadboard](https://github.com/rwaldron/johnny-five/tree/master/docs/breadboard). These are ideally named such that they match the JS program file. 
- Add a new "topic" to `tpl/programs.json` that will be dedicated to your new additions. Take a look at the [Intel Edison + Grove IoT Kit](https://github.com/rwaldron/johnny-five/blob/master/tpl/programs.json#L1553-L1575) for guidance.


I realize this is a lot of work, but here are the benefits: 

- Your work will be made directly available to users of both this repo and the [johnny-five.io](http://johnny-five.io/examples/) website
  - This makes it a linkable/sharable "go to guide". For example, I can easily link to the [Intel Edison + Grove IoT Kit](http://johnny-five.io/examples/#intel-edison--grove-iot-kit) examples. The website will be undergoing updates soon, which will make this more appealing ;)
  - The next site iteration will also link [platforms](http://johnny-five.io/platform-support/) to platform-specific guides like this.
- Your work will be featured/highlighted in a [news post](http://johnny-five.io/news/)


In short, you'll be providing an awesome community resource :)

**Comment by warycat** [comment]
I will spend some time on fritzing diagrams. Unfortunately I burned my intel edison with the mini breakout board. I am waiting for my tessel2 to come. I will work on it this weekend.

**Comment by warycat** [comment]
In docs/breadboard, there are two types of files. Can the project build the website with .fzz and render them in .png?

**Comment by dtex** [comment]
@warycat No, you have to export the png's from Fritzing

**Comment by rwaldron** [comment]
There is still a bit of work to do on the source in these files. Since the actual files are all in a directory that's unknown to the linter config, they aren't being checked—which is giving us a false "pass" on travis. I'm going to update the config so that it will look in all subdirectories of `eg/`, then you can do: 

```
# this only needs to be done once
git remote add upstream https://github.com/rwaldron/johnny-five.git
```

Then...

```
git pull --rebase upstream master
```

Then...

```
grunt
```

**Comment by rwaldron** [comment]
I've made it (hopefully) a bit easier now, so you can do this: 


```
# this only needs to be done once
git remote add upstream https://github.com/rwaldron/johnny-five.git
```

Then...

```
git pull --rebase upstream master
```

Then...

```
grunt qc:examples
```


Refs
- https://github.com/rwaldron/johnny-five/commit/71d5aca9bde3353e201b785d70665d4ac9583a6e
- https://github.com/rwaldron/johnny-five/commit/5a455c74610b4f3b5bbf8d0e3baa91b94e86e208

**Comment by warycat** [comment]
It's a little bit tedious to write es5. And I try tessel2 the other day. It does not support npm install on board.  The uploading is really slow compared to edison. I can mount edison as a disk and edit the code and run node-supervisor on board.

**Comment by rwaldron** [comment]
> It's a little bit tedious to write es5. 

I didn't ask you to write ES5? I just asked you to lint the code :\

> And I try tessel2 the other day. It does not support npm install on board. 

It doesn't need to, npm install locally and your project is bundled and deployed

> The uploading is really slow compared to edison. 

If you were uploading local code to an Edison, the process would be exactly the same and take the same time. 

> I can mount edison as a disk and edit the code and run node-supervisor on board.

That's cool, I've used `rsub` with Edison and like that as well.

**Comment by warycat** [comment]
I do grunt qc:examples with no errors, but grunt test is breaking.

File: eg/TheArduinoStarterKit/ColorMixingLamp.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  var led = new five.Led.RGB([6,5,3]);
+  var b = new Buffer(3);
+  var sensors = new five.Sensors([{pin:"A0"},{pin:"A1"},{pin:"A2"}]);
+  sensors.on("change",function() {
+    b[0] = sensors[0].analog;
+    b[1] = sensors[1].analog;
+    b[2] = sensors[2].analog;
+    var c = b.toString("hex");
+    console.log(c);
+    led.color("#" + c);
```
Comment (rwaldron) [inline eg/TheArduinoStarterKit/ColorMixingLamp.js]: If you're interested, this can be simplified. The `color` method accepts an array of 8-bit values, so this works as well: 

```js
var sensors = new five.Sensors(["A0", "A1", "A2"]);
sensors.on("change", function() {
  led.color([
    sensors[0].analog, 
    sensors[0].analog,
    sensors[0].analog, 
  ]);
});
```


Or even...


```js
var sensors = new five.Sensors(["A0", "A1", "A2"]);
sensors.on("change", function() {
  led.color(
    sensors.map(function(sensor) {
      return sensor.value >> 2;
    })
  );
});
```

File: eg/TheArduinoStarterKit/CrystalBall.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  var button = new five.Button(6);
+  var lcd = new five.LCD({
+    pins: [12, 11, 5, 4, 3, 2],
+  });
+  var crystalBall = new CrystalBall(lcd);
+  crystalBall.ask();
+  button.on("press",function() {
+    crystalBall.reply();
+  });
+});
```
Comment (rwaldron) [inline eg/TheArduinoStarterKit/CrystalBall.js]: Very cool

File: eg/TheArduinoStarterKit/DigitalHourglass.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+var five = require("johnny-five");
+var Edison = require("edison-io");
+var board = new five.Board({io: new Edison()});
+
+board.on("ready", function() {
+  var leds = new five.Leds([2,3,4,5,6,7]);
+  var tilt = new five.Sensor({pin:8,type:"digital"});
```
Comment (rwaldron) [inline eg/TheArduinoStarterKit/DigitalHourglass.js]: Can also be written as: 

```js
five.Sensor.Digital(8);
```

But I will leave that to you

File: eg/TheArduinoStarterKit/KnockLock.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  var yellow = new five.Led(3);
+  var green = new five.Led(4);
+  var red = new five.Led(5);
+  var servo = new five.Servo(9);
+
+  var box = new Box(servo, red, green, yellow);
+
+});
```
Comment (rwaldron) [inline eg/TheArduinoStarterKit/KnockLock.js]: I'm making this tomorrow ;)

**Comment by rwaldron** [comment]
I love how much you've used the collection classes, very cool :)


I will see what I can do about getting this to pass for you tomorrow. Thanks again for all your work on this :D

File: eg/TheArduinoStarterKit/ColorMixingLamp.js (hunk 1 above)
Comment (warycat) [inline eg/TheArduinoStarterKit/ColorMixingLamp.js]: how about this?  
sensors.on("change",function() {
    var c = sensors.reduce(function(s, sensor) {
      return s + sensor.analog.toString(16);
    }, "#");
    console.log(c);
    led.color(c);
  });
